File: frontend/src/types/activity.ts
/**
 * Activity Types for Slot-Level Schedule Events
 *
 * Activities are distinct from Rotations:
 * - Rotation = Multi-week block assignment (e.g., "Neurology" for 4 weeks)
 * - Activity = Slot-level event (e.g., "FM Clinic AM", "LEC Wednesday PM")
 */

import type { UUID } from './api';

// ============================================================================
// Activity Category
// ============================================================================

/**
 * Categories for activities to enable filtering and business logic.
 */
export type ActivityCategory = 'clinical' | 'educational' | 'administrative' | 'time_off';

/**
 * All valid activity categories.
 */
export const ACTIVITY_CATEGORIES: ActivityCategory[] = [
  'clinical',
  'educational',
  'administrative',
  'time_off',
];

/**
 * Human-readable labels for activity categories.
 */
export const ACTIVITY_CATEGORY_LABELS: Record<ActivityCategory, string> = {
  clinical: 'Clinical',
  educational: 'Educational',
  administrative: 'Administrative',
  time_off: 'Time Off',
};

// ============================================================================
// Activity
// ============================================================================

/**
 * Represents a slot-level schedule activity.
 *
 * @example
 * ```typescript
 * const fmClinic: Activity = {
 *   id: 'uuid',
 *   name: 'FM Clinic',
 *   code: 'fm_clinic',
 *   displayAbbreviation: 'C',
 *   activityCategory: 'clinical',
 *   isProtected: false,
 *   displayOrder: 1,
 * };
 * ```
 */
export interface Activity {
  /** Unique identifier */
  id: UUID;
  /** Human-readable name (e.g., 'FM Clinic', 'Lecture') */
  name: string;
  /** Stable identifier for solver (e.g., 'fm_clinic') */
  code: string;
  /** Short code for UI grid (e.g., 'C', 'LEC') */
  displayAbbreviation: string | null;
  /** Category: clinical, educational, administrative, timeOff */
  activityCategory: ActivityCategory;
  /** Tailwind color class for text */
  fontColor: string | null;
  /** Tailwind color class for background */
  backgroundColor: string | null;
  /** ACGME supervision requirement */
  requiresSupervision: boolean;
  /** True for locked slots (e.g., LEC Wednesday PM) */
  isProtected: boolean;
  /** ACGME clinical hour limit tracking */
  countsTowardClinicalHours: boolean;
  /** Sort order for UI */
  displayOrder: number;
  /** Whether this activity is archived (soft deleted) */
  isArchived: boolean;
  /** When the activity was archived */
  archivedAt: string | null;
  /** When the activity was created */
  createdAt: string;
  /** When the activity was last updated */
  updatedAt: string;
}

/**
 * Request payload for creating an activity.
 */
export interface ActivityCreateRequest {
  name: string;
  code: string;
  displayAbbreviation?: string | null;
  activityCategory: ActivityCategory;
  fontColor?: string | null;
  backgroundColor?: string | null;
  requiresSupervision?: boolean;
  isProtected?: boolean;
  countsTowardClinicalHours?: boolean;
  displayOrder?: number;
}

/**
 * Request payload for updating an activity.
 */
export interface ActivityUpdateRequest {
  name?: string;
  code?: string;
  displayAbbreviation?: string | null;
  activityCategory?: ActivityCategory;
  fontColor?: string | null;
  backgroundColor?: string | null;
  requiresSupervision?: boolean;
  isProtected?: boolean;
  countsTowardClinicalHours?: boolean;
  displayOrder?: number;
}

/**
 * Response from activity list API.
 */
export interface ActivityListResponse {
  activities: Activity[];
  total: number;
}

// ============================================================================
// Activity Requirement (Soft Constraints)
// ============================================================================

/**
 * Week numbers within a 4-week block.
 * null means the requirement applies to all weeks.
 */
export type ApplicableWeeks = number[] | null;

/**
 * Represents a soft constraint for how many half-days of an activity
 * should be scheduled within a rotation.
 *
 * @example
 * ```typescript
 * const fmClinicReq: ActivityRequirement = {
 *   id: 'uuid',
 *   rotationTemplateId: 'rotation-uuid',
 *   activityId: 'fm-clinic-uuid',
 *   activity: fmClinicActivity,
 *   minHalfdays: 3,
 *   maxHalfdays: 5,
 *   targetHalfdays: 4,
 *   applicableWeeks: null, // all weeks
 *   preferFullDays: true,
 *   preferredDays: [1, 2, 3, 4, 5], // Mon-Fri
 *   avoidDays: [0, 6], // Sun, Sat
 *   priority: 75,
 * };
 * ```
 */
export interface ActivityRequirement {
  /** Unique identifier */
  id: UUID;
  /** The rotation template this requirement belongs to */
  rotationTemplateId: UUID;
  /** The activity this requirement is for */
  activityId: UUID;
  /** Nested activity data */
  activity: Activity;
  /** Minimum half-days required (hard floor) */
  minHalfdays: number;
  /** Maximum half-days allowed (hard ceiling) */
  maxHalfdays: number;
  /** Preferred count (soft optimization target) */
  targetHalfdays: number | null;
  /** Week numbers [1,2,3,4] or null for all weeks */
  applicableWeeks: ApplicableWeeks;
  /** Prefer AM+PM together for same activity */
  preferFullDays: boolean;
  /** Preferred day numbers [0-6, 0=Sun, 6=Sat] */
  preferredDays: number[] | null;
  /** Days to avoid [0-6, 0=Sun, 6=Sat] */
  avoidDays: number[] | null;
  /** Priority 0-100, higher = more important */
  priority: number;
  /** When the requirement was created */
  createdAt: string;
  /** When the requirement was last updated */
  updatedAt: string;
}

/**
 * Request payload for creating an activity requirement.
 */
export interface ActivityRequirementCreateRequest {
  activityId: UUID;
  minHalfdays?: number;
  maxHalfdays?: number;
  targetHalfdays?: number | null;
  applicableWeeks?: ApplicableWeeks;
  preferFullDays?: boolean;
  preferredDays?: number[] | null;
  avoidDays?: number[] | null;
  priority?: number;
}

/**
 * Request payload for updating an activity requirement.
 */
export interface ActivityRequirementUpdateRequest {
  minHalfdays?: number;
  maxHalfdays?: number;
  targetHalfdays?: number | null;
  applicableWeeks?: ApplicableWeeks;
  preferFullDays?: boolean;
  preferredDays?: number[] | null;
  avoidDays?: number[] | null;
  priority?: number;
}

/**
 * Request payload for bulk updating all requirements for a rotation.
 */
export interface ActivityRequirementBulkUpdateRequest {
  requirements: ActivityRequirementCreateRequest[];
}

/**
 * Response from activity requirements list API.
 */
export interface ActivityRequirementListResponse {
  requirements: ActivityRequirement[];
  totalHalfdays: number;
}

// ============================================================================
// Priority Levels
// ============================================================================

/**
 * Standard priority levels for activity requirements.
 * Higher = more important to the solver.
 */
export const PRIORITY_LEVELS = {
  /** Nice to have (0-30) */
  LOW: { min: 0, max: 30, label: 'Low' },
  /** Should satisfy (31-60) */
  MEDIUM: { min: 31, max: 60, label: 'Medium' },
  /** Strong preference (61-90) */
  HIGH: { min: 61, max: 90, label: 'High' },
  /** Near-hard constraint (91-100) */
  CRITICAL: { min: 91, max: 100, label: 'Critical' },
} as const;

/**
 * Get the priority level label for a numeric priority.
 */
export function getPriorityLabel(priority: number): string {
  if (priority <= 30) return PRIORITY_LEVELS.LOW.label;
  if (priority <= 60) return PRIORITY_LEVELS.MEDIUM.label;
  if (priority <= 90) return PRIORITY_LEVELS.HIGH.label;
  return PRIORITY_LEVELS.CRITICAL.label;
}

/**
 * Get the color class for a priority level (for UI badges).
 */
export function getPriorityColor(priority: number): string {
  if (priority <= 30) return 'bg-gray-100 text-gray-700';
  if (priority <= 60) return 'bg-yellow-100 text-yellow-700';
  if (priority <= 90) return 'bg-orange-100 text-orange-700';
  return 'bg-red-100 text-red-700';
}

// ============================================================================
// Utility Functions
// ============================================================================

/**
 * Formats applicable weeks for display.
 *
 * @example
 * formatApplicableWeeks(null) // "All weeks"
 * formatApplicableWeeks([1, 2, 3]) // "Weeks 1-3"
 * formatApplicableWeeks([1, 3, 4]) // "Weeks 1, 3, 4"
 */
export function formatApplicableWeeks(weeks: ApplicableWeeks): string {
  if (!weeks || weeks.length === 0) return 'All weeks';
  if (weeks.length === 4) return 'All weeks';

  const sorted = [...weeks].sort((a, b) => a - b);

  // Check for consecutive sequence
  const isConsecutive = sorted.every((w, i) => i === 0 || w === sorted[i - 1] + 1);

  if (isConsecutive && sorted.length > 1) {
    return `Weeks ${sorted[0]}-${sorted[sorted.length - 1]}`;
  }

  return `Weeks ${sorted.join(', ')}`;
}

/**
 * Formats a day number to a short name.
 * 0 = Sun, 1 = Mon, ..., 6 = Sat
 */
export function formatDayShort(day: number): string {
  const days = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
  return days[day] ?? '???';
}

/**
 * Formats preferred/avoid days for display.
 */
export function formatDays(days: number[] | null): string {
  if (!days || days.length === 0) return 'None';
  return days.map(formatDayShort).join(', ');
}

File: backend/app/schemas/activity.py
"""Activity schemas for slot-level schedule events.

Activities are distinct from Rotations:
- Rotation = Multi-week block assignment (e.g., "Neurology" for 4 weeks)
- Activity = Slot-level event (e.g., "FM Clinic AM", "LEC Wednesday PM")
"""

from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator


class ActivityCategory(str, Enum):
    """Categories for activities to enable filtering and business logic."""

    CLINICAL = "clinical"
    EDUCATIONAL = "educational"
    ADMINISTRATIVE = "administrative"
    TIME_OFF = "time_off"


# Valid categories as tuple for validation
VALID_ACTIVITY_CATEGORIES = tuple(c.value for c in ActivityCategory)


class ActivityBase(BaseModel):
    """Base activity schema with shared fields."""

    name: str = Field(
        ..., description="Human-readable name (e.g., 'FM Clinic', 'Lecture')"
    )
    code: str = Field(
        ...,
        max_length=50,
        description="Stable identifier for solver (e.g., 'fm_clinic')",
    )
    display_abbreviation: str | None = Field(
        None,
        max_length=20,
        description="Short code for UI grid (e.g., 'C', 'LEC')",
    )
    activity_category: str = Field(
        ...,
        description="Category: clinical, educational, administrative, time_off",
    )
    procedure_id: UUID | None = Field(
        None,
        description="Optional procedure ID for credentialed activities (e.g., VAS, SM)",
    )
    font_color: str | None = Field(None, description="Tailwind color class for text")
    background_color: str | None = Field(
        None, description="Tailwind color class for background"
    )
    requires_supervision: bool = Field(
        True, description="ACGME supervision requirement"
    )
    is_protected: bool = Field(False, description="True for locked slots (e.g., LEC)")
    counts_toward_clinical_hours: bool = Field(
        True, description="ACGME clinical hour limit"
    )
    display_order: int = Field(0, description="Sort order for UI")
    capacity_units: int | None = Field(
        None,
        ge=0,
        description="FMC physical capacity units consumed by this activity (0 for non-physical)",
    )

    @field_validator("name")
    @classmethod
    def validate_name(cls, v: str) -> str:
        """Validate name is not empty."""
        if not v or not v.strip():
            raise ValueError("name cannot be empty")
        return v.strip()

    @field_validator("code")
    @classmethod
    def validate_code(cls, v: str) -> str:
        """Validate code format."""
        if not v or not v.strip():
            raise ValueError("code cannot be empty")
        v = v.strip().lower()
        # Code should be lowercase alphanumeric with underscores or hyphens
        # Hyphens needed for codes like LV-PM, C-AM, W-AM, FMIT-R
        if not all(c.isalnum() or c in "_-" for c in v):
            raise ValueError(
                "code must be lowercase alphanumeric with underscores or hyphens only"
            )
        return v

    @field_validator("activity_category")
    @classmethod
    def validate_category(cls, v: str) -> str:
        """Validate activity category."""
        if v not in VALID_ACTIVITY_CATEGORIES:
            raise ValueError(
                f"activity_category must be one of: {VALID_ACTIVITY_CATEGORIES}"
            )
        return v


class ActivityCreate(ActivityBase):
    """Schema for creating a new activity."""

    pass


class ActivityUpdate(BaseModel):
    """Schema for updating an activity. All fields optional."""

    name: str | None = None
    code: str | None = None
    display_abbreviation: str | None = None
    activity_category: str | None = None
    font_color: str | None = None
    background_color: str | None = None
    requires_supervision: bool | None = None
    is_protected: bool | None = None
    counts_toward_clinical_hours: bool | None = None
    display_order: int | None = None
    capacity_units: int | None = None
    procedure_id: UUID | None = None

    @field_validator("code")
    @classmethod
    def validate_code(cls, v: str | None) -> str | None:
        """Validate code format if provided."""
        if v is None:
            return v
        v = v.strip().lower()
        # Hyphens needed for codes like LV-PM, C-AM, W-AM, FMIT-R
        if not all(c.isalnum() or c in "_-" for c in v):
            raise ValueError(
                "code must be lowercase alphanumeric with underscores or hyphens only"
            )
        return v

    @field_validator("activity_category")
    @classmethod
    def validate_category(cls, v: str | None) -> str | None:
        """Validate activity category if provided."""
        if v is None:
            return v
        if v not in VALID_ACTIVITY_CATEGORIES:
            raise ValueError(
                f"activity_category must be one of: {VALID_ACTIVITY_CATEGORIES}"
            )
        return v


class ActivityResponse(ActivityBase):
    """Schema for activity response with all fields."""

    id: UUID
    is_archived: bool = False
    archived_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ActivityListResponse(BaseModel):
    """Schema for list of activities."""

    activities: list[ActivityResponse]
    total: int


# ============================================================================
# Activity Requirement Schemas
# ============================================================================


class ActivityRequirementBase(BaseModel):
    """Base schema for activity requirements."""

    activity_id: UUID = Field(..., description="The activity this requirement is for")
    min_halfdays: int = Field(0, ge=0, le=28, description="Minimum half-days required")
    max_halfdays: int = Field(14, ge=0, le=28, description="Maximum half-days allowed")
    target_halfdays: int | None = Field(
        None, ge=0, le=28, description="Preferred count"
    )
    applicable_weeks: list[int] | None = Field(
        None,
        description="Week numbers [1,2,3,4] or null for all weeks",
    )
    prefer_full_days: bool = Field(True, description="Prefer AM+PM together")
    preferred_days: list[int] | None = Field(
        None,
        description="Preferred day numbers [1-5 for Mon-Fri]",
    )
    avoid_days: list[int] | None = Field(
        None,
        description="Days to avoid [0,6 for Sun,Sat]",
    )
    priority: int = Field(
        50, ge=0, le=100, description="0-100, higher = more important"
    )

    @field_validator("applicable_weeks")
    @classmethod
    def validate_weeks(cls, v: list[int] | None) -> list[int] | None:
        """Validate week numbers are 1-4."""
        if v is None:
            return v
        for week in v:
            if week < 1 or week > 4:
                raise ValueError("Week numbers must be 1-4")
        return sorted(set(v))  # Remove duplicates and sort

    @field_validator("preferred_days", "avoid_days")
    @classmethod
    def validate_days(cls, v: list[int] | None) -> list[int] | None:
        """Validate day numbers are 0-6."""
        if v is None:
            return v
        for day in v:
            if day < 0 or day > 6:
                raise ValueError("Day numbers must be 0-6 (0=Sun, 6=Sat)")
        return sorted(set(v))


class ActivityRequirementCreate(ActivityRequirementBase):
    """Schema for creating an activity requirement."""

    pass


class ActivityRequirementUpdate(BaseModel):
    """Schema for updating an activity requirement."""

    min_halfdays: int | None = Field(None, ge=0, le=28)
    max_halfdays: int | None = Field(None, ge=0, le=28)
    target_halfdays: int | None = Field(None, ge=0, le=28)
    applicable_weeks: list[int] | None = None
    prefer_full_days: bool | None = None
    preferred_days: list[int] | None = None
    avoid_days: list[int] | None = None
    priority: int | None = Field(None, ge=0, le=100)


class ActivityRequirementResponse(ActivityRequirementBase):
    """Schema for activity requirement response."""

    id: UUID
    rotation_template_id: UUID
    activity: ActivityResponse  # Nested activity data
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ActivityRequirementListResponse(BaseModel):
    """Schema for list of activity requirements."""

    requirements: list[ActivityRequirementResponse]
    total_halfdays: int  # Sum of target_halfdays across all requirements


class ActivityRequirementBulkUpdate(BaseModel):
    """Schema for bulk updating activity requirements."""

    requirements: list[ActivityRequirementCreate]
